### PR TITLE
Improvements to ADPCM-B implementation

### PR DESCRIPTION
**Pull request by aaronsgiles**
Specific fixes:
* Implement EOS and PLAYING flags as sticky bits that require manual clearing on all implementations.
* Implement correct end-of-sample behavior, skipping final 3 nibbles.
* Implement read access patterns that match hardware tests.
* Match behavior of "dummy" reads to hardware tests.
* Hook up verified edge case behaviors for over-reading and over-writing data.
* Add framework for running ADPCM tests to compare against hardware results.

### CODE REVIEW

**Comment by aaronsgiles** [comment]
Continuing conversation from https://github.com/aaronsgiles/ymfm/pull/38 over here.

File: src/ymfm_adpcm.cpp [hunk 1]
```diff
@@ -613,68 +611,101 @@ void adpcm_b_channel::write(uint32_t regnum, uint8_t value)
...
-			else
+			// advance; if we hit the end, signal EOS and put ourselves back in the latching state
+			// also turn off record mode, as subsequent writes are ignored
+			if (advance_address())
 			{
-				m_owner.intf().ymfm_external_write(ACCESS_ADPCM_B, m_curaddress++, value);
-				m_status = STATUS_BRDY;
+				set_reset_status(STATUS_EOS);
+				m_curaddress = LATCH_ADDRESS;
+				m_regs.write(0x00, m_regs.read(0x00) & ~0x40);
```
Comment (hyano) [inline src/ymfm_adpcm.cpp:693]: This behavior is seen only when REPEAT=1.
When REPEAT=0, the second cycle is also a write operation (record mode).

```
if (m_regs.repeat())
    m_regs.write(0x00, m_regs.read(0x00) & ~0x40);
```

File: src/ymfm_adpcm.cpp [hunk 2]
```diff
@@ -613,68 +611,101 @@ void adpcm_b_channel::write(uint32_t regnum, uint8_t value)
...
 			}
 		}
+
+		// writes in external non-record mode appear to behave like a read in that it will advance
+		// the address and consume a nibble, but the last written value will still be present; note
+		// that this code runs even after the end of a normal write that sets the EOS bit
+		if (m_regs.external())
```
Comment (hyano) [inline src/ymfm_adpcm.cpp:700]: Is checking !m_regs.record() required?

```
if (m_regs.external() && !m_regs.record())
{
```

**Review by hyano — COMMENTED** [review]
The latest adpcm-test code works for my project.

For unusual cases, I noticed two points that seem to be incorrect.

File: src/ymfm_adpcm.cpp (hunk 1 above)
Comment (aaronsgiles) [reply src/ymfm_adpcm.cpp:693]: Hmmm, so if REPEAT=0 then what happens? Does the write continue at the start address? Did you have a test for that case?

Comment (hyano) [reply src/ymfm_adpcm.cpp]: > Hmmm, so if REPEAT=0 then what happens? Does the write continue at the start address?

Yes.

> Did you have a test for that case?

One of the examples is:
https://github.com/hyano/opna-analyze/blob/main/test/mem_rw.py
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/mem_rw.txt

I've converted to your C++ framework. Please check it.
[mem_rw.cpp.txt](https://github.com/aaronsgiles/ymfm/files/9159676/mem_rw.cpp.txt)

Currently, I cannot understand actual meaning of REPEAT bit.

Comment (aaronsgiles) [reply src/ymfm_adpcm.cpp]: Ok, I have checked in your rw test. I notice that the previous work to fix dummy reads broke one of the tests, so I made some more changes to fix that. 

I also made another change to initialize memory to 0x80 because the mem_rw test causes a read from uninitialized memory at the end of the "READ ADDRESS 0fff- CHANGE START(10/RESET)" test, so that the dummy reads in the next text "READ ADDRESS 0fff- CHANGE STOP" return that uninitialized memory. Your results showed 0x80 there, so I initialized my memory to 0x80 so that we'd match, but it would be better if the test were fixed.

Comment (hyano) [reply src/ymfm_adpcm.cpp]: > Your results showed 0x80 there, so I initialized my memory to 0x80 so that we'd match, but it would be better if the test were fixed.

Yes, I also noticed it.
When I create mem_rw.py, dummy read mechanism was unclear and I ignored it.

I have updated my test to fill extended area. And confirmed reading expected values.
https://github.com/hyano/opna-analyze/blob/main/test/mem_rw.py
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/mem_rw.txt

The log and signals are also recaptured, so you can refer if needed.
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/

Thank you for using my tests even if it is very ad-hoc and experimental...

[mem_rw.cpp.txt](https://github.com/aaronsgiles/ymfm/files/9172712/mem_rw.cpp.txt)

**Comment by hyano** [comment]
I've added test for another unusual cases which READ & REC=1.
* https://github.com/hyano/opna-analyze/blob/main/test/mem_rw16.py
* https://github.com/hyano/opna-analyze/blob/main/log/mem_rw16_mden/mem_rw16.txt

In that condition, I observed:
* No memory access happens (confirmed by MDEN signal).
* The internal data buffer becomes empty after reading 2bytes and BRDY becomes 0.

It would be a hint for behavior of internal data buffer.

[mem_rw16.cpp.txt](https://github.com/aaronsgiles/ymfm/files/9172804/mem_rw16.cpp.txt)

**Comment by happppp** [comment]
If we don't find a bad regression, the ADPCM changes have been cherry picked for MAME 0.284: https://github.com/mamedev/mame/commit/66deabe1768df8e0ee3767688a8bfec32df27f72

**Comment by aaronsgiles** [comment]
@happppp Is this code change still active as-is in MAME? If so, I'll incorporate it now that it's had some decent testing.

**Comment by happppp** [comment]
The emulation related changes: yes, no known regressions so far.
The test files: no, I did not copy those over.

**Comment by aaronsgiles** [comment]
Ok thanks for confirming.
